// node-sass-magic-importer/dist/toolbox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const glob = require("glob");
const path = require("path");
const resolve = require("resolve");
const build_include_paths_1 = require("./functions/build-include-paths");
const clean_import_url_1 = require("./functions/clean-import-url");
const escape_selector_1 = require("./functions/escape-selector");
const parse_node_filters_1 = require("./functions/parse-node-filters");
const parse_selector_filters_1 = require("./functions/parse-selector-filters");
const process_raw_selector_filters_1 = require("./functions/process-raw-selector-filters");
const resolve_glob_url_1 = require("./functions/resolve-glob-url");
const resolve_package_key_1 = require("./functions/resolve-package-key");
const resolve_package_url_1 = require("./functions/resolve-package-url");
const resolve_url_1 = require("./functions/resolve-url");
const sass_glob_pattern_1 = require("./functions/sass-glob-pattern");
const sass_url_variants_1 = require("./functions/sass-url-variants");
const split_selector_filter_1 = require("./functions/split-selector-filter");
exports.buildIncludePaths = build_include_paths_1.buildIncludePathsFactory(path);
exports.cleanImportUrl = clean_import_url_1.cleanImportUrlFactory();
exports.escapeSelector = escape_selector_1.escapeSelectorFactory();
exports.parseNodeFilters = parse_node_filters_1.parseNodeFiltersFactory();
exports.processRawSelectorFilters = process_raw_selector_filters_1.processRawSelectorFiltersFactory(exports.escapeSelector);
exports.resolvePackageKey = resolve_package_key_1.resolvePackageKeyFactory();
exports.sassGlobPattern = sass_glob_pattern_1.sassGlobPatternFactory(path);
exports.sassUrlVariants = sass_url_variants_1.sassUrlVariantsFactory(path);
exports.resolveGlobUrl = resolve_glob_url_1.resolveGlobUrlFactory(glob, path);
exports.resolvePackageUrl = resolve_package_url_1.resolvePackageUrlFactory(resolve, exports.resolvePackageKey, exports.sassUrlVariants);
exports.resolveUrl = resolve_url_1.resolveUrlFactory(glob, path, exports.sassGlobPattern);
exports.splitSelectorFilter = split_selector_filter_1.splitSelectorFilterFactory();
exports.parseSelectorFilters = parse_selector_filters_1.parseSelectorFiltersFactory(exports.processRawSelectorFilters, exports.splitSelectorFilter);
